common/scala/src/main/scala/whisk/core/database/CouchDBStoreProvider.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.database
+
+import akka.actor.ActorSystem
+import spray.json.RootJsonFormat
+import whisk.common.Logging
+import whisk.core.WhiskConfig
+import whisk.spi.SpiModule
+
+/**
+ * A CouchDB implementation of ArtifactStoreProvider
+ */
+class CouchDBStoreProvider extends ArtifactStoreProvider {
+    def makeStore[D <: DocumentSerializer](config: WhiskConfig, name: WhiskConfig => String)(
+        implicit jsonFormat: RootJsonFormat[D],
+        actorSystem: ActorSystem,
+        logging: Logging): ArtifactStore[D] = {
+        require(config != null && config.isValid, "config is undefined or not valid")
+        require(config.dbProvider == "Cloudant" || config.dbProvider == "CouchDB", "Unsupported db.provider: " + config.dbProvider)
+        assume(Set(config.dbProtocol, config.dbHost, config.dbPort, config.dbUsername, config.dbPassword, name(config)).forall(_.nonEmpty), "At least one expected property is missing")
+
+        new CouchDbRestStore[D](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, name(config))
+    }
+}
+
+class CouchDBStoreProviderModule extends SpiModule[ArtifactStoreProvider] {
+    def getInstance = new CouchDBStoreProvider
+}
```

Comment:
Which interface? The `SpiModule[Spi]` interface makes it so that a) SPIs are constructed in a specific way (currently using ServiceLoader), and (via ServiceLoader) all impls are listed in a specific file (META-INF/services/whisk.spi.SpiModule). Otherwise, we have to provide a separate/custom mechanism for discovering impls of a specific interface. ServiceLoader works fine for this IMHO. It also provides an abstraction point to allow either DI or some extended dependency support in the future, without changing the signature of the SPI itself.